design: BGP FRR integration

markdgray opened the pull request:
This design document discusses integration of FRR as an
alternative BGP agent implementation for MetalLB.

Signed-off-by: Mark Gray <mark.d.gray@redhat.com>

champtar commented on design/0001-frr.md:
```diff
@@ -0,0 +1,347 @@
+
+##### speakerlist
+
+speakerlist uses the vagrant memberlist package to create a cluster of bgp
```
s/vagrant/Hashicorp/

markdgray replied:
> Hi @markdgray
> 
> I went through the document quickly, for me this integration is a great occasion to create a separate BGP controller instead of a keeping the idea of BGP speaker around:

I'm not sure I understand. Are you suggesting creating another Kubernetes controller just to configure BGP? MetalLB would then use that to configure FRR indirectly? I did consider this model but I have some concern about the latency of receiving and handling changes to kubernetes objects by this controller. I am unsure how responsive we would need to be?
> 
>     1. you don't need to run FRR/BGP on each nodes
> 
>     2. you don't need hostNetwork=true
> 
>     3. you don't need the added Linux capabilities
> 
> 
> Having some BGP controllers running only on the control-plane would be great for security
> 
> My 2 cents

champtar replied:
Not sure to understand the concern about responsiveness, the current speakers are using client-go to talk to the API, such BGP controller would do the same. It's just 2 different models, having all nodes peer to the router is a bit weird to me, it does allow you to run BFD which is nice, but I don't know if it's standard / expected to peer all nodes or not.

markdgray replied:
Ok, I think I see what you are saying. You are suggesting that we don't need all nodes to peer to routers. There could be efficiencies to be gained by only having a subset to peer to routers. That may be possible. I think we would still need to have metallb instances on each node to detect when a node goes down but, potentially, we could not deploy FRR to every node. That would preclude BFD as you mentioned which would be a disadvantage. 

I don't _think_ we could always guarantee there would only be one BGP controller as it may not always be possible to route BGP messages to some routers from a particular node (there could be some administrative constraint). That is why I said "subset" above. If this is the case, it would complicate the design as you would need to figure out which nodes should have an instance of FRR and which shouldn't. Maybe some one else has an opinion?

markdgray replied:
Also, this makes the assumption that the BGP controller can advertise routes with a NEXT_HOP that is not itself.

markdgray replied:
I did a bit of research on this. Typically, a router will host its own BGP stack which will then peer with a BGP agent on another router. In our case, the node is acting as a router and the BGP speaker is acting as the router's BGP stack. Therefore, I think the model currently employed by MetalLB is the more standard approach. Additionally, as you suggested, we will need to deploy FRR (or an alternative) like this in order to make use of BFD. Therefore, I think we should maintain the current model. What do you think?

russellb replied:
I agree with your conclusion, Mark

champtar replied:
Thinking about it more, and mostly to have BFD, the current approach is the way to go. Deploying on each node to use memberlist as a BFD replacement just to have less peering makes little sense.
We could just add the option in the Chart to disable L2 and in that case remove the extra capabilities, so people running BGP only can make it a bit more secure, but that's out of scope for this PR

markdgray replied:
Ok, thanks for the review @champtar

champtar commented:
Hi @markdgray 

I went through the document quickly, for me this integration is a great occasion to create a separate BGP controller instead of a keeping the idea of BGP speaker around:
1. you don't need to run FRR/BGP on each nodes
2. you don't need hostNetwork=true
3. you don't need the added Linux capabilities

Having some BGP controllers running only on the control-plane would be great for security

My 2 cents

msherif1234 commented on design/0001-frr.md:
```diff
@@ -0,0 +1,347 @@
+  deployment at cluster installation time.
+* Ensure feature parity with current BGP implementation.
+* Reinstate thorough BGP testing infrastructure.
+
```
FRR BGP configuration has more capabilities and options than what MetalLB BGP support today for example VRF support how do you plan to extend MetalLB to take advantage of those features ?

markdgray replied:
@msherif1234 That could be a future enhancement but this enhancement just deals with achieving feature parity with the current implementation. All testing will reflect that.

markdgray replied:
@msherif1234 I will add a non-goal to explicitly call this out.

msherif1234 commented on design/0001-frr.md:
```diff
@@ -0,0 +1,347 @@
+FRR will be deployed as a container. The speaker application will be configured
+at startup to use the FRR bgp daemon and will provide details for connection to
+the daemon (e.g. location of a socket)
+
```
FRR will be deployed as daemonSet in the same name space as metallb-system right ?

markdgray replied:
Yes, I think that would be the case.

markdgray replied:
@msherif1234 I changed the test to suggest this.

msherif1234 commented on design/0001-frr.md:
```diff
@@ -0,0 +1,347 @@
+integration with the “metrics” struct which updates prometheus in response to
+BGP events. It may be necessary to refactor these types of functionality into a
+separate package for reuse amongst BGP implementations.
+
```
what about exising limitation betweem MetallB BGP and any external component that can also do BGP peering like the following 
- https://metallb.universe.tf/configuration/calico/
- https://metallb.universe.tf/configuration/kube-router/
will moving to FRR help fix the above issues/limitations ?

markdgray replied:
If the current native BGP agent peers with an external router, this will preclude another BGP agent on that node peering with the same router. This is due to BGP’s conflict resolution algorithm as stated in the links above. This is a limitation with the native BGP implementation in MetalLB.

At an architectural level, allowing MetalLB to reuse/share a BGP agent that is already present on the node would appear to resolve this issue as we would let the BGP agent resolve conflicts and act as the single point of contact with the peered router. However, there may be implementation details that would prevent this. Also, the ability to share the BGP agent is currently not an explicit goal of this enhancement. I think that as development of this feature continues, it should be something to investigate and I will update this enhancement accordingly. Thanks @msherif1234!

msherif1234 commented on design/0001-frr.md:
```diff
@@ -0,0 +1,347 @@
+
+The "speaker" application supports a Layer 2 mode and a BGP mode. The following
+discussion focuses on BGP mode.
+
```
do we need to have debugging/troubleshooting section and explain how can we use FRR BGP debugging ?

markdgray replied:
I don't think it is needed for this enhancement but I think it should be added to the implementation. I will try to add something to the enhancement to remind us to add that documentation to the implementation.

markdgray replied:
I added this to the Goals section.

msherif1234 commented on design/0001-frr.md:
```diff
@@ -0,0 +1,347 @@
+the daemon (e.g. location of a socket)
+
+The container will expose a control interface that will initially be used to
+configure FRR via the `vtysh` application and eventually via its gRPC golang
```
are u planning to use vtysh to configure FRR or the standard MetalLB configmap and in future CRD ?
if we will use k8s way to configure do we need to disable vtysh to change the config ?

markdgray replied:
I have no plans to allow direct configuration of FRR via the MetalLB ConfigMap (or CRD).  There may be use-cases in which configuration of the FRR instance will need to be shared between two actors (MetalLB and another) but in that case we will partition FRR appropriately.

russellb replied:
In fact, I'd call out keeping FRR as an implementation detail as an explicit design goal.  What do you think, Mark?

mccv1r0 commented on design/0001-frr.md:
```diff
@@ -0,0 +1,347 @@
+* The purpose of this is not to provide an interface to plug any BGP agent into
+  MetalLB. Although, this may provide a useful template for others to do this.
+* This will not support runtime changing of BGP agent type.
+
```
Non-Goal?  * Allowing anything else beside MetalLB to use FRR?

Or is it a goal (in MetalLB repo??) to let something else e.g. HAProxy/nginx/whatever inform FRR about prefix(es) to advertise as well?

markdgray replied:
Yeah, this is a good addition and msherif1234 mentioned something similar above. I think it would be worth adding it as a goal and spending some time ensuring that this can be done. Thanks.

mccv1r0 replied:
Could you then run FRR w/out MetalLB?   Just use e.g. HAProxy or nothing... just advertise prefixes?

russellb replied:
You could, but you still need something that understands what to advertise, and where to advertise it from.  That's the layer that MetalLB solves for the LoadBalancer Service use case.

markdgray replied:
@mccv1r0 @russellb I have been thinking about this some more. I think it might be better to not allow other things to advertise prefixes - at least by default - as this will open the door to unusual configurations that may introduce weird bugs that will be difficult for the upstream community to support and debug. It could potentially make an unfriendly "out-of-the-box experience" if a user attempted to do this. As such I think this should be a _non-goal_.

However, I think this is an important use case and may be a common deployment model. Therefore, we need to consider this during development in order to not preclude the ability for a _smart_ consumer of MetalLB to be able to easily engineer this if required. It could also be that, with time, we do allow this capability. What do you think?

russellb commented on design/0001-frr.md:
```diff
@@ -0,0 +1,347 @@
+## Summary
+
+The purpose of this enhancement is to use [Free Range
+Routing](https://frrouting.org/) (FRR) to provide BGP functionality in MetalLB.
```
```suggestion
Routing](https://frrouting.org/) (FRR) as an alternative to provide BGP functionality in MetalLB.
```

markdgray replied:
Reworded this.

markdgray commented:
I have added a section to discuss evaluation of alternative routing stacks and the motivation for selecting FRR.

s1061123 commented on design/0001-frr.md:
```diff
@@ -0,0 +1,391 @@
+configure FRR via the `vtysh` application and eventually via its gRPC golang
+bindings. This will be used to implement bgp.New() and bgp.Close() for session
+creation and bgp.Set() for prefix advertisement.
+
```
Do you have any idea to gather bgpd logging? At least, for troubleshooting, we need to gather log from bgpd somehow and provide user to look the logs. One candidate could be stdout and user can see through `kubectl logs` command...

markdgray replied:
Yeah, I think this would be the natural way to do this. We can configure FRR using something like `log stdout LEVEL`.

s1061123 replied:
Shall we note it in this doc? or it is just taken care of implementation?

russellb commented on design/0001-frr.md:
```diff
@@ -0,0 +1,402 @@
+  configured with a gRPC interface which can generate Go bindings.
+* Extensibility options: FRR provides additional extensibility via its `dplane`
+  and `FPM` interfaces. This may ease integration with Kubernetes networking
+  providers. There does not appear to be equivalents in BIRD.
```
It could be worth elaborating on this a bit to give an example of what type of integration you have in mind here?

markdgray replied:
Sure, done.

russellb commented on design/0001-frr.md:
```diff
@@ -0,0 +1,402 @@
+* As FRR implements each routing protocol as a seperate executable, the
+  footprint of the container can potentially be reduced in the FRR case by not
+  packaging unneeded routing protocols. BIRD allows disabling protocols
+  via a configuration option.
```
Is this a prediction, or were the sizes checked?

russellb replied:
Not necessarily ... it's interesting if confirmed and the size difference is significant, but if we aren't sure then I don't think it helps make the case.  I guess I'm saying I'd back it up with some analysis or just remove it.

markdgray replied:
removed.

russellb commented on design/0001-frr.md:
```diff
@@ -0,0 +1,402 @@
+  provides support for additional routing protocols which, although not
+  strictly required for the MetalLB use case, may be useful in deployments
+  where a user has modified MetalLB to share the routing stack with other
+  applications (not just MetalLB).
```
We also discussed in another thread making such modifications a non-goal.  We should probably clarify this somehow.  Maybe it's that we don't want to support it directly, but also not prevent it if someone where to inject custom configuration into their frr container?

On another note, support of OSPF is also in a long standing issue and should be considered in scope for MetalLB in the future.  https://github.com/metallb/metallb/issues/256

johananl commented on design/0001-frr.md:
```diff
@@ -0,0 +1,411 @@
+
+This would couple MetalLB with the FRR community so that it would be somewhat
+dependent on this community for bug-fixes and further feature improvements.
+However, there would be the native BGP agent as a fallback for cluster
+administrators that did not want to deploy FRR.
```
How would we handle cases where an upstream change to FRR is required by MetalLB but the FRR maintainers are either slow to implement it or they decide not to do that at all for whatever reason? And what about e.g. urgent security fixes?

- We can't simply fall back to the native stack because FRR has features the native stack doesn't.
- It doesn't make sense to ensure we only add FRR features that are supported natively because getting features "for free" was a central motivation for introducing FRR in the first place.
- If we end up adding functionality to MetalLB to keep up with relevant FRR features, perhaps it doesn't make sense to direct out efforts towards FRR in the first place.

I think that we can consider switching to FRR as the default BGP stack after some time has passed and we've gained confidence in the project, its release cadence etc. If we do that, maybe the problems above would be addressed.

markdgray replied:
As FRR is an external dependency, I don't see a straightforward way to achieve this. A lot depends on the vibrancy of the project and the community as you have suggested.

From what I have seen/read, it does appear to be very active project under the umbrella of the Linux Foundation with regular releases and commits. The development workflow is documented [here](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#development-release-cycle) and does specify their process for back ports and release branches. Regarding security, they have a dedicated `security@lists.frrouting.org` address for handling security issues.

However, I suspect that any issues that we find would still be very relevant to their other use cases as what we are doing (compared to what other BGP deployments with FRR would be doing) is relatively straightforward. FWIW, their BGP PR/issue backlog seems quite active (https://github.com/FRRouting/frr/labels/bgp)

markdgray replied:
I will also remove the native implementation as a fallback as I think we are now talking about its eventually retirement.

johananl commented on design/0001-frr.md:
```diff
@@ -0,0 +1,411 @@
+with external routers, this may invoke BGP’s conflict resolution algorithm. This
+is detailed [here](https://metallb.universe.tf/configuration/calico/) for the
+Calico case and [here](https://metallb.universe.tf/configuration/kube-router/)
+for the Kube-Router case. This enhancement may be a step towards resolving these
```
In what way does the FRR integration address e.g. the MetalLB + Calico problem? Not sure I got that. Could you elaborate?

markdgray replied:
I think I see your confusion and I will update. From my understanding of the problem, only one agent can peer with the remote. Therefore, if MetalLB's native implementation is pairing with the remote, no other BGP software can establish a session. As the native implementation is currently limited and is tightly coupled with MetalLB so it could not be shared with another entity wishing to advertise with the remote. This enhancement improves this situation because FRR could potentially be shared by another entity allowing other prefixes to be advertised via this session. In the Calico case, this would require plugging in their own (BIRD?) BGP implementation as an alternative MetalLB BGP implementation which could then be reused by Calico.

Jean-Daniel replied:
Sharing the same FRR instance may create other challenges like detecting conflicting configurations and also properly tracking the advertised routes owner (to make sure a component does not delete routes from an other component, and to properly detect staling routes)

markdgray replied:
Yes, I agree but I am just specifying here that this is a "step towards" resolving these issues as it should provide a way to eventually solve this. I will add this clarification. Is that OK?

johananl commented on design/0001-frr.md:
```diff
@@ -0,0 +1,411 @@
+## Motivation
+
+The motivation for this enhancement is to provide an alternative
+production-ready BGP agent for use in MetalLB. Overall, this should reduce the
+maintenance cost for the MetalLB project by reducing the effort required for
+adding additional BGP features and bug fixes. For example, there are a number of
```
I'd say the BGP package has been the most stable area of the speaker since I joined the project. Therefore, I'm not sure we can argue that switching to FRR would reduce the project's maintenance burden.

To counter the above, I'd say we (the maintainers) have been slow in adding e.g. new BGP features to MetalLB, which IMO gives merit to the idea of getting such features "for free" by adopting FRR.

markdgray replied:
Ok, I will clarify that in the document.

johananl commented on design/0001-frr.md:
```diff
@@ -0,0 +1,411 @@
+
+* Provide a configuration option to allow use of FRR as a BGP agent in a MetalLB
+  deployment at cluster installation time.
+* Ensure feature parity with current BGP implementation.
```
What is our long-term vision in terms of feature parity between the two BGP implementations? Do we aim to retire the native BGP stack (e.g. after we gain confidence with FRR)?
I can already see a fairly immediate scenario where we want to expose a new knob in the config which turns on a feature which exists in FRR but not in the native BGP stack. I'd say we need a strategy for handling that...

markdgray replied:
I will address this by adding a paragraph describing the roadmap towards this retirement.

johananl commented on design/0001-frr.md:
```diff
@@ -0,0 +1,411 @@
+BGP. There is an experimental gRPC interface. This interface may need to be
+productionized through the FRR community. When this has been satisfactorily
+achieved, we can start [Story 3](#story-3). Until that time, in order to
+mitigate this risk, we can wrap the FRR ‘vtysh’ command line interface in Go
```
Does wrapping `vtysh` entail automating an interactive CLI using Go code? This sounds fragile to me. I'd opt for updating a config file if FRR supports live config reloading until we have e.g. gRPC support.

markdgray replied:
I wasn't considering wrapping the interactive `vtysh` shell but rather send commands using vtysh which should be less fragile. i.e. `vtysh -c "command"`.

Using a config file is an option. There are some limitations with that that would be preferable to address (https://github.com/FRRouting/frr/issues/2128)

markdgray replied:
Shall I update the document to state the preferred method (e.g. `vtysh -c cmd`)?

rata replied:
LGTM. @johananl ?

johananl reviewed (commented):
Hey there @markdgray.

First of all, let me say this is an excellent design proposal :clap: It's readable, well-rationalized and makes a lot of sense. I'd say we can use it as a template for future proposals.

Second, your overview of the MetalLB codebase looks great. I'd consider converting this part of the proposal to a developer-oriented document. Would love to get a PR for that if you're up for it.

Overall I like the proposal a lot. I do have some concerns and reservations which I've expressed as inline comments. My biggest concern is the lack of strategy around long-term maintenance of the BGP stacks. My experience tells me we will very quickly abandon the old stack because FRR has more features compared to the native stack. This could be a good thing, but I think we should be more explicit about our intentions here.

Happy to hear your thoughts, and would love to see more from you in this project :slightly_smiling_face:

johananl commented on design/0001-frr.md:
```diff
@@ -0,0 +1,411 @@
+component. This is because FRR will only peer with BGP agents outside of the
+cluster and not with each other.
+
+## Drawbacks
```
Another drawback IMO is expanding the speaker's attack surface. In terms of security I'd say it's a plus to have a very thin BGP implementation. For example, you can't confuse MetalLB by sending malicious advertisements to the speaker because it doesn't support *accepting* BGP updates, only sending them.

We *may* be able to mitigate this using FRR config hardening, I'm not familiar with FRR to know for sure. Anyway, I'd say this is a point to consider.

markdgray replied:
Good point, I will update and add some examples configurations that I can find.

markdgray commented:
> Hey there @markdgray.
> 
> First of all, let me say this is an excellent design proposal clap It's readable, well-rationalized and makes a lot of sense. I'd say we can use it as a template for future proposals.

Great!

> 
> Second, your overview of the MetalLB codebase looks great. I'd consider converting this part of the proposal to a developer-oriented document. Would love to get a PR for that if you're up for it.
> 

I was originally planning something like that so I can refactor the document as you have suggested.

> Overall I like the proposal a lot. I do have some concerns and reservations which I've expressed as inline comments. My biggest concern is the lack of strategy around long-term maintenance of the BGP stacks. My experience tells me we will very quickly abandon the old stack because FRR has more features compared to the native stack. This could be a good thing, but I think we should be more explicit about our intentions here.
> 

I will look through these comments and address them in the next couple of days or maybe next week.

> Happy to hear your thoughts, and would love to see more from you in this project slightly_smiling_face

Thanks @johananl !

markdgray commented on design/0001-frr.md:
```diff
@@ -0,0 +1,411 @@
+
+#### FRR deployment
+
+FRR will be deployed as a container. The current proposal is to deploy it
```
I had been thinking about deploying it as a separate pod as it will 1) make it easier to independently manage the lifecycle of MetalLB and 2) make it easier for a user to use their own MetalLB pod. However, I don't really have any strong convictions about this at the moment.

Regarding communication, `vtysh` (the FRR command line tool) communicates with the daemons using a unix socket (I need to confirm if there is a TCP socket option). The gRPC interface is over a TCP socket.

russellb replied:
Can you describe the security of the vtysh interface if they're running in separate pods?  A benefit of sticking them in the same pod is a simpler security model for the communication.

markdgray replied:
> A nice side effect of running in a separate pod would be that the metallb speaker would no longer require host network access. Only the FRR pod would be on the host network.

Yes, good point. I will add this point in the next revision.

markdgray replied:
> Can you describe the security of the vtysh interface if they're running in separate pods? A benefit of sticking them in the same pod is a simpler security model for the communication.

Another good point. I did some investigation (and I am not an expert). vtysh is available over unix sockets or tcp sockets. 

For unix sockets, the socket is protected by unix file permissions.

For tcp sockets, it is protected by a password that is specified in the config file. It should be noted, on further investigation, that for tcp sockets you need to open a telnet session to a particular port and that is now the remote connectivity can be achieved. There is no native support for something like SSL that I can see. You can also limit what IP address that listens for connections (e.g. 127.0.0.1 or 0.0.0.0, etc)

russellb replied:
I really like the separate lifecycle management benefit in particular.  I've also seen requirements for "TLS everywhere" before though, so this could come back to bite us.  Even if it's localhost only, it still doesn't seem secure enough.

My feeling right now is to stick it within the same pod to keep an acceptable security model.  Hopefully we can do something better in the future with the grpc interface?  Unfortunately from a quick look, the grpc interface doesn't support anything better.  We may have an easier time sticking a secure proxy in front of it with grpc.  In any case, if we need to improve this in frr, putting that energy on the grpc side is probably a better investment.

markdgray replied:
Yes, I agree that this is something to flag to the FRR community.

On the community call this week, we discussed this. The general consensus was that we should use the TCP/IP interface on 127.0.0.1 and deploy as a container in an MetalLB pod. This reduces the complexity from a Metal LB project perspective. If a consumer wants to deploy FRR as a pod they could then (somehow) change the IP address used to connect to FRR.

champtar replied:
Is it possible to use a socket file ?

markdgray replied:
Socket file is another supported communication method.

russellb replied:
I think this settled on running it within the speaker pod, right?

markdgray replied:
I think so. I will summarize this conversation and commit an update.

champtar commented:
I haven't taken the time to read the proposal in full, so sorry if the response in in the proposal (just did some quick search).
One possible behavior change is that current BGP code doesn't listen on port 179 on the host, making it possible to run along side calico in BGP mode (default if I remember correctly).
Will FRR listen on port 179 ?

markdgray commented on design/0001-frr.md:
```diff
@@ -0,0 +1,411 @@
+  interface between this DaemonSet and MetalLB. Therefore, if any non-backwards
+  compatible changes are made in this interface, this will need to be managed in
+  the MetalLB ‘speaker’ component.
+* Upgrade/downgrade will remove a node from the MetalLB memberlist by unlabeling
```
That is a little confusing but looking again I see that sList is not passed to the BGP controller. Re-reading the code again, I think the general principal still applies. We can remove all sessions from the node using the nodeSelector mechanism and then, after the upgrade, we can re-add the sessions to the node using the nodeSelector.

One thing I am not sure about is the following. If Node A's BGP sessions are brought down while we update FRR, do we want that to be detected as a dead peer. If so, that will force all advertisements to be redirected to another node. Is that OK?

markdgray commented:
> I haven't taken the time to read the proposal in full, so sorry if the response in in the proposal (just did some quick search).
> One possible behavior change is that current BGP code doesn't listen on port 179 on the host, making it possible to run along side calico in BGP mode (default if I remember correctly).
> Will FRR listen on port 179 ?

It can be configured (https://docs.frrouting.org/en/latest/bgp.html#starting-bgp). It looks like that is a configuration parameter in MetalLB so we would just need to implement that?

markdgray commented:
> Second, your overview of the MetalLB codebase looks great. I'd consider converting this part of the proposal to a developer-oriented document. Would love to get a PR for that if you're up for it.

I added this part as a separate file. Will I split it out into a separate file?

russellb commented:
Overall I'm quite happy with this and am ready to give my +1.  I'd like to resolve the remaining open question on whether FRR would run within the metallb speaker pod or not, and if it's separate, how we would secure that communication.

russellb reviewed (approved):
+1 from me.

We should get an explicit ack from @johananl @rata @daxmc99 as well

markdgray commented:
> You could also mention the security concerns

done

briantopping commented:
This is a really great proposal, appreciate the attention to why FRR was chosen over BIRD. 

Also seems like a great first use of feature gates if something like that could be adopted: #867

rata reviewed (commented):
@markdgray Thanks, I hope this really help to expand MetalLB protocol support! Sorry for the late review.

I'm not a fan of adding a new BGP-only dependency (as this helped to highlight, it creates overhead for layer 2 in this extra unneeded container, need to release on FRR security issues) but if no one has the time to add more features/protocols to the built-in implementation, I think this step is inevitable. So, let's just find the least painful way to do it and move forward :)

Added some comments, but seems like a nice improvement overall!

@champtar Hm, interesting... Not sure I follow, though. Can you please elaborate?

> I went through the document quickly, for me this integration is a great occasion to create a separate BGP controller instead of a keeping the idea of BGP speaker around:
> 
>     1. you don't need to run FRR/BGP on each nodes

Sometimes you do, though. It setups like Equinix Metal, you most likely need that. I think it is quite common to need it for BGP.

>     2. you don't need hostNetwork=true

You mean if the node SNATs the outbound connection? Because the BGP peer surely needs to answer TCP packets to the pod :)

I think I'm missing something you say, though. But if one mostly needs to run BGP speakers on every node, what you suggests still applies?

Jean-Daniel commented:
> I'm not a fan of adding a new BGP-only dependency (as this helped to highlight, it creates overhead for layer 2 in this extra unneeded container, need to release on FRR security issues) but if no one has the time to add more features/protocols to the built-in implementation, I think this step is inevitable. So, let's just find the least painful way to do it and move forward :)

I'm not sure the time to add BGP features (IPV6, BFD, …) and maintain them will be less than the time needed to make FRR container conditional in case it is really needed, especially with the helm based deployment  ;-).

uablrek commented:
At Ericsson we use the metallb `controller` but have developed an own `speaker` with the functions we need, for example;

* IPv6
* BFD
* Static routes (with BFD)
* Traffic separation (separate peers)

At a meeting with `RedHat` we were asked to contribute with our speaker configuration as input to this PR, perhaps even as a definition of done;

* [ECFE_Configuration_Examples.md](https://github.com/metallb/metallb/files/6631459/ECFE_Configuration_Examples.md)
* [ECFE_Configuration_Examples.pdf](https://github.com/metallb/metallb/files/6631462/ECFE_Configuration_Examples.pdf)

markdgray commented on design/0001-frr.md:
```diff
@@ -0,0 +1,340 @@
+
+#### FRR deployment
+
+FRR will be deployed in a container as part of the speaker Pod. This will
```
* FRR does provide official container images which I hope we could reuse: https://hub.docker.com/r/frrouting/frr
* FRR supports "releases" (http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#releases) every 4 months which may be minor or major releases. Security and bug fixes are backported up to a year after the release but may also be contributed after that date. There are some discussions about LTS releases but nothing concrete at this point.

Not sure if there is anything you would like me to update in the doc to reflect this comment?

rata replied:
Cool, thank you very much! No need to add it, if it becomes a question others have we can add it for simplicity, but as you prefer.

We might need to take it into account in the future, like we might need to release MetalLB to adjust to some FRR things, but nothing to do in the short term AFAICS

markdgray commented on design/0001-frr.md:
```diff
@@ -0,0 +1,340 @@
+  specific FRR bugs without upgrading MetalLB.
+* This enhancement assumes that each node’s MetalLB ‘speaker’ component can be
+  stopped (for an arbitrary time) and restarted without disruption of MetalLB as
+  a whole (To be confirmed)
```
> At the MetalLB layer you can restart speaker pods with no "global" issue, just two things come to mind:

Yes, this is what I was referring to and thanks for pointing out the link as it makes the behavior clear. Therefore, for upgrade there should not be a significant issue with stopping speaker pods to carry out the upgrade considering the potential mitigations available. Also, I presume, this is similar to current behavior during upgrade. If this is the case, I will remove the "To be confirmed" and add the link you shared?

rata replied:
Yes, exactly, it is the current behavior on upgrades too. Sure, update with that link, thanks!

markdgray commented on design/0001-frr.md:
```diff
@@ -0,0 +1,340 @@
+tests to deal with an upgrade to a newer FRR version.
+
+* We will allow independent upgrade of the FRR component in order to resolve any
+  specific FRR bugs without upgrading MetalLB.
```
> Ah, ok, so the idea is that the end-user upgrades the FRR container on his own? 
I am unsure exactly what you are asking but the following two comments should answer: 

As the container is part of the MetalLB pod
* we will need to restart the pod on container upgrade
* we will need to define the version of the container in the pod yaml

> Does upstream provides images on each release, do you plan to use frr from some distro or how one would upgrade exactly?

* Yes. I was assuming the version would be specified in the pod yaml but maybe we could make it configurable? wdyt?

rata replied:
Sorry, not sure I follow the diff between "frr version in the pod yaml" vs "configurable". You mean to have some helm knob? That will be only for helm users, right?

I think configurable in the pod yaml (as any version of any container, nothing special) seems simpler. But I think I'm missing what you are comparing this to :)

markdgray replied:
@rata I guess I mean there is a default version of the FRR pod but we would provide it as a configurable option. Maybe via helm. Is this ok?

markdgray replied:
Updated the text here.

markdgray commented on design/0001-frr.md:
```diff
@@ -0,0 +1,340 @@
+
+### Non-Goals
+
+* Coexistence of two types of BGP agents (e.g. MetalLB native BGP agent AND FRR
+  BGP agent) is not a goal of this work.
```
How about:

`The same type of BGP agent (e.g. MetalLB native BGP agent or FRR BGP agent) MUST be used cluster-wide. Inter-mixing of different BGP agent types on node or within the cluster is not guaranteed to work or be tested.`

rata replied:
LGTM, thanks!

markdgray commented on design/0001-frr.md:
```diff
@@ -0,0 +1,340 @@
+  the cluster maintains the same version or a compatible version of FRR.
+  However, it is required that each version of FRR is compatible with the BGP
+  agent with which it is peering. After FRR is updated, the ‘speaker’ component
+  can be restarted and the node can be labelled again.
```
Yes, I think so. I will update.

markdgray commented on design/0001-frr.md:
```diff
@@ -0,0 +1,340 @@
+BGP. There is an experimental gRPC interface. This interface may need to be
+productionized through the FRR community. When this has been satisfactorily
+achieved, we can start [Story 3](#story-3). Until that time, in order to
+mitigate this risk, we can wrap the FRR ‘vtysh’ command line interface in Go
```
I'm not sure. I do agree with you that gut-feeling would be that it is stable as it would need to be scriptable. However, I would concede that this is a (low) risk item.

rata replied:
Cool, thanks!

markdgray commented on design/0001-frr.md:
```diff
@@ -0,0 +1,340 @@
+
+#### FRR deployment
+
+FRR will be deployed in a container as part of the speaker Pod. This will
```
I am not sure what the resource utilization is but I can't expect it would be to high. Therefore I would prefer on all setups.

When idle, you can see the utilization here which is very low:
```
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
frr      2186725  0.0  0.0 329756  5728 ?        S<sl 06:36   0:00 /usr/lib/frr/zebra -d -A 127.0.0.1
frr      2186735  0.0  0.0 259736  7532 ?        S<sl 06:36   0:00 /usr/lib/frr/bgpd -d -A 127.0.0.1
root     2186749  0.0  0.0  80824  3448 ?        S<s  06:36   0:00 /usr/lib/frr/watchfrr -d -r /usr/lib/frr/frr restart %s -s /usr/lib/frr/frr start %s -k /usr/lib/frr/frr stop %s zebra bgpd
```

rata replied:
Ok, let's go with that and we can adjust if needed. Let's hope the requests for cpu/mem for that container don't need to be too big (not for the idle case you show, for the case where it is indeed used).

markdgray commented on design/0001-frr.md:
```diff
@@ -0,0 +1,340 @@
+
+#### FRR deployment
+
+FRR will be deployed in a container as part of the speaker Pod. This will
```
Yes, the BGP port is configurable so we could change it in that case: https://docs.frrouting.org/en/latest/bgp.html#cmdoption-bgpd-p

rata replied:
Cool, thanks!

markdgray commented on design/0001-frr.md:
```diff
@@ -0,0 +1,340 @@
+which may be required at a later stage.
+
+Initially, control of the FRR container will be achieved by sending messages to
+localhost over the `vtysh` TCP socket, protected by a password. This will go
```
This was discussed in a community call IIRC. The use case would be a deployment that already has an FRR instance deployed somewhere and wanted to reuse that instance (for example sharing it with another application). This is something we would like to do at Red Hat in the future. 

We can specify port and address to listen on 

```
       -A, --vty_addr vty-addr
              Specify an IP/IPv6 address to bind the TCP VTY interface to.  It is generally recommended to specify ::1 or 127.0.0.1.  For reasons of backwards compatibility, the default is to listen on all interfaces.
```

```
  -P, --vty_port vty-port
              Override the daemon's default TCP VTY port (each daemon has a different default value upwards of 2600, listed below.)  Specifying 0 disables the TCP VTY interface.

```

markdgray replied:
> Can frr listen only on localhost? Just curious: why password protected? And do you envision such password as shared across all metallb downloads or unique per installation?
Just added security. Probably unique per node.

rata replied:
(the reply quoted your answer too)

Thanks! Ok, if that is useful at Red Hat, seems better to use that on all cases rather than having a config knob just for that :)

Can you please mention the password will be unique per node? It is a detail I'd like to remember when reviewing the code, so better if it is explicit :)

markdgray replied:
Done

markdgray commented on design/0001-frr.md:
```diff
@@ -0,0 +1,340 @@
+relevant Kubernetes e2e tests (or custom MetalLB tests) against this test
+cluster. Further investigation will be required as part of [Story 1](#story-1)
+in order to determine the relevance of any Kubernetes e2e tests and if any other
+e2e tests will need to be developed.
```
I will update this as I phrased it badly. I meant we could reuse some code from the Kubernetes e2etest but not necessarily the tests themselves. I have already implemented a smoke test https://github.com/metallb/metallb/commit/bacd8ac5a4eedb2502e414a8418c22fce321b76e that should be a good start.

rata replied:
Cool, thanks!

markdgray commented on design/0001-frr.md:
```diff
@@ -0,0 +1,340 @@
+  providers. There does not appear to be equivalents in BIRD. This would
+  enable FRR to integrate with dataplanes other than the Linux kernel. For
+  example, 1) hardware dataplanes 2) an Open Flow configured dataplane like Open
+  vSwitch 3) DPDK dataplane.
```
Open Flow in particular is interesting to Red Hat due to ovn-kubernetes. I guess the others are theoretical but it wouldn't surprise me if they became relevant in the future.

rata replied:
Oh, cool, thanks! And they do not apply to bird too (or you don't know at least), I guess?

markdgray replied:
I reworded this slightly. They main point here is FRR provides more options for future extensibility if required. Hence, is more flexible.

markdgray commented on design/0001-frr.md:
```diff
@@ -0,0 +1,340 @@
+  applications (not just MetalLB). Although sharing the routing stack with
+  other applications is an unsupported, non-goal of this enhancement, some users
+  may want to inject custom configurations into the BGP stack. [OSPF](https://github.com/metallb/metallb/issues/256)
+  is one example of an additional protocol requested by the MetalLB community.
```
I have done an evaluation and FRR does appear to support more. Some of which may be more relevant/important than others. ISIS and l2vpn are two examples. It's worth reminding that there is nothing in this PR that precludes BIRD for a future enhancement. In fact, we have explicitly stated a goal that this will allow different BGP agents to be plugged in.

rata replied:
Ok, can we mention these two that bird does not provide, then? I think that has "historical value" when us (or others) in the future conside other options :)

markdgray replied:
Done

markdgray commented on design/0001-frr.md:
```diff
@@ -0,0 +1,340 @@
+  whereas FRR is dual-licensed under [GPL2](https://github.com/FRRouting/frr/blob/master/COPYING)
+  and [LGPL2.1](https://github.com/FRRouting/frr/blob/master/COPYING-LGPLv2.1).
+  LGPL2.1 will give us more flexibility in terms of how we link with the FRR
+  application.
```
If we link it will require GPL'ing our code. However, if we don't link - it shouldn't be an issue. Therefore, LGPL is more flexible for us.

rata replied:
Sorry, brain fart I _think_ (but with license, I'm never sure when I'm getting it right :)). FRR is dual licensed, so we can indeed use it as GPL2, link with it with the go bindings in the future and no need to GPL our code, right?

I think the CNCF requires our code is licensed under Apache 2 or something like that.

markdgray replied:
CNCF seems to state that our code should be Apache as stated here: https://github.com/cncf/foundation/blob/master/charter.md#11-ip-policy.

LGPL license allows linking **without** requiring you to GPL your own code. For example, libc is LGPL (https://en.wikipedia.org/wiki/GNU_C_Library). 

GPL would require us to change our license to GPL if we linked with it which would be unacceptable.

Therefore, I believe dual-licensed [FRR](https://github.com/FRRouting/frr/blob/master/COPYING)[FRR](https://github.com/FRRouting/frr/blob/master/COPYING-LGPLv2.1) is more flexible versus single-licensed [GPL](https://gitlab.nic.cz/labs/bird/-/blob/master/README) BIRD.

I don't think I need to update anything here?

markdgray commented on design/speaker.md:
```diff
@@ -0,0 +1,128 @@
+as part of the Metallb build process.
+
+The "speaker" application supports a Layer 2 mode and a BGP mode. The Layer 2
+mode advertises prefixes using Gratuitous ARP, whereas the BGP mode
```
I will clarify this in the next revision.

markdgray commented:
@rata Thanks for the comprehensive review. I had a few follow-up questions before I do another revision of this PR.

rata commented:
@markdgray I think I answer them all, thanks again!

@uablrek Cool. Not sure how to read your message, though: do you mean you would prefer to contribute your speaker implementation instead of using FRR? Or are you saying that MetalLB going to FRR is beneficial and you'll be able to drop the custom speaker and use metallb upstream speaker implementation?

russellb commented:
> @uablrek Cool. Not sure how to read your message, though: do you mean you would prefer to contribute your speaker implementation instead of using FRR? Or are you saying that MetalLB going to FRR is beneficial and you'll be able to drop the custom speaker and use metallb upstream speaker implementation?

Based on some offline conversation, I think this is an expression of support of the general direction we're taking. They have an alternative implementation of it, but are not able to open source it at this time (and maybe that could change in the future). In the meantime, their configuration examples offer some inspiration for further extensions we may want to make for additional features they find important in their own implementation.  In other words, take it as input for some future target features in the implementation.

markdgray commented:
Is there any consensus? I think everything has been addressed.

rata commented:
I don't think I'll have time to have another look soon, but don't blovk the PR on my final ACK. It was quite close to ready IMHO when I reviewed the last time, also :)

Thanks again for the PR!

johananl commented on design/speaker.md:
```diff
@@ -0,0 +1,128 @@
+
+speakerlist uses the Hashicorp memberlist package to create a cluster of
+speakers, manage membership of this cluster and to detect member failures using
+the “gossip” protocol. This is only used in the speaker's L2 mode.
```
```suggestion
a gossip protocol. This is only used in the speaker's L2 mode.
```

"Gossip protocols" is a category of protocols rather than the name of a specific protocol.

johananl commented on design/speaker.md:
```diff
@@ -0,0 +1,128 @@
+### main
+
+The main package contains the main() entrypoint for the speaker application. It
+creates an L2 and/or BGP controller, a speakerList and a k8s client. The k8s
```
IMO we should use either "Layer 2" or "L2" consistently throughout the document.

markdgray replied:
Changed to "Layer 2"

johananl commented on design/speaker.md:
```diff
@@ -0,0 +1,128 @@
+
+### config
+
+Parses configmap into an internal representation of the configuration described
```
- Missing article before "configmap".
- Let's use either "config map" or "configmap" consistently.

markdgray replied:
I added the article and changed to "ConfigMap" as this is the proper type of the object.

johananl commented on design/speaker.md:
```diff
@@ -0,0 +1,128 @@
+
+### k8s
+
+Kubernetes client/watcher
```
```suggestion
A Kubernetes client/watcher.
```

johananl commented on design/speaker.md:
```diff
@@ -0,0 +1,128 @@
+
+## BGP session creation
+
+New BGP sessions are established by calling newBGP() in the speaker "main"
```
```suggestion
New BGP sessions are established by calling newBGP() in the speaker's "main"
```

johananl commented on design/speaker.md:
```diff
@@ -0,0 +1,128 @@
+## BGP session creation
+
+New BGP sessions are established by calling newBGP() in the speaker "main"
+package which calls bgp.New() returning a session interface. The session
```
```suggestion
package which calls bgp.New() which in turn returns a session interface. The session
```

johananl reviewed (commented):
Thanks a lot for the effort @markdgray.

Please don't be overwhelmed by the number of comments :slightly_smiling_face: These are mostly cosmetic nitpicks. I took the time to add them since I believe being lenient about details in one part of a codebase has a tendency to lower the quality bar for other areas (such as actual code), too. Please treat all the language/phrasing/capitalization-related comments as non blockers.

Following is a summary of the only "big" concerns I have left (more info inline):

- Configuring FRR *imperatively* using `vtysh` poses IMO a risk of running into unexpected state problems between MetalLB and FRR.
- I'm worried about the emphasis on managing "external" agents in this proposal.
- Are we good to "embed" FRR inside MetalLB in terms of OSS licenses?

russellb commented on design/0001-frr.md:
```diff
@@ -0,0 +1,341 @@
+only one agent on a node may peer with a remote BGP agent. This is detailed
+[here](https://metallb.universe.tf/configuration/calico/) for the Calico case
+and [here](https://metallb.universe.tf/configuration/kube-router/) for the
+Kube-Router case. This enhancement may be a step towards resolving these
```
That's fair. It's a bit more of a speculative benefit, and not one provided directly by the proposed change.  Removing mention of it seems OK to me.

markdgray replied:
Ok, I will remove this section and another reference to it further down (i.e. `and, although not
  strictly required for the MetalLB use case, may be useful in deployments where
  a user has modified MetalLB to share the routing stack with other applications
  (not just MetalLB). Although sharing the routing stack with other applications
  is an unsupported, non-goal of this enhancement, some users may want to inject
  custom configurations into the BGP implementation.`

russellb commented on design/0001-frr.md:
```diff
@@ -0,0 +1,341 @@
+  custom configurations into the BGP stack. [OSPF](https://github.com/metallb/metallb/issues/256)
+  is one example of an additional protocol requested by the MetalLB community
+  that is provided by BIRD and FRR.
+* Licensing: BIRD is licensed under [GPL2](https://github.com/BIRD/bird)
```
In short, yes, what we're doing as proposed is fine.

We aren't actually embedding it.  The only "embedding" is a reference to its container image.

The most relevant case is a potential future where we may want to use a library from FRR to interact with it via its new northbound API.  AIUI, we are OK if that library is LGPL licensed.  When we get to that point, I'm happy to get a more official opinion on that matter to help with any concerns.  Even if FRR were to license their library in an incompatible way, the API is a grpc API and we could write our own license-compatible interface if necessary.  Again though, this isn't applicable to the current plan though.

markdgray replied:
Agreed. I don't think there is any issue here.

johananl replied:
Thanks for the elaborate explanation!

markdgray commented on design/0001-frr.md:
```diff
@@ -0,0 +1,341 @@
+The purpose of this enhancement is to use [Free Range
+Routing](https://frrouting.org/) (FRR) as an alternative BGP implementation
+in MetalLB. When directed to, MetalLB will publish prefixes via FRR rather than
+MetalLB’s current built-in BGP speaker.
```
Ok, I will use "implementation"

markdgray replied:
I have updated, and we should be using (consistently) the term "BGP implementation" to refer to the implementation of BGP being used by MetalLB. 

I also use the term "peer" or "BGP peer" to refer to a remote BGP stack that we are peering with.

I use the term "BGP software" to refer to other BGP software that could be running on the node (but not part of MetalLB BGP)

markdgray commented on design/0001-frr.md:
```diff
@@ -0,0 +1,341 @@
+#### Story 1
+
+As a developer, I want to be able to run a comprehensive set of BGP system
+tests, preferably based on kubernetes e2e tests, in order to ensure the MetalLB
```
I think I got them all now. Thanks.

markdgray commented on design/0001-frr.md:
```diff
@@ -0,0 +1,341 @@
+  specific FRR bugs without upgrading MetalLB. MetalLB will ship with a default
+  version but it will be possible to configure the version somehow (e.g. through
+  helm).
+* At the MetalLB layer you can restart speaker pods with no "global" issue.
```
Done and checked across the whole commit.

markdgray commented on design/0001-frr.md:
```diff
@@ -0,0 +1,341 @@
+Deploying FRR as a seperate Pod was also considered and it would give some
+advantages, such as the removal of the requirement for host networking from the
+speaker Pod, and the seperation of the FRR and speaker Pod lifecycles but
+would add complexity to the end-user.
```
fixed and checked across commit.

markdgray commented on design/0001-frr.md:
```diff
@@ -0,0 +1,341 @@
+## Drawbacks
+
+* This would couple MetalLB with the FRR community so that it would be somewhat
+  dependent on this community for bug-fixes and further feature improvements.
```
made consistent (i.e. bug fix or bug fixes)

markdgray commented on design/0001-frr.md:
```diff
@@ -0,0 +1,341 @@
+   implementation but does not add any additional features.
+3. Depending on the successful outcome of (2), raise a deprecation notice on
+   the MetalLB website, in the repository, and on the Slack channel indicating
+   that the native BGP implmentation will be deprecated and no future
```
feature :D .. Fixed

markdgray commented on design/0001-frr.md:
```diff
@@ -0,0 +1,341 @@
+This configuration interface will be used by MetalLB to implement bgp.New() and
+bgp.Close() for session creation and bgp.Set() for prefix advertisement.
+
+#### others
```
Yes, I have made it clearer now.

markdgray commented on design/0001-frr.md:
```diff
@@ -0,0 +1,341 @@
+## Motivation
+
+The motivation for this enhancement is to provide an alternative
+production-ready BGP agent for use in MetalLB. Overall, this should reduce the
```
updated as per previous comment.

markdgray commented on design/0001-frr.md:
```diff
@@ -0,0 +1,341 @@
+### Goals
+
+* Provide a configuration option to allow use of alternate BGP agents in a
+  MetalLB deployment at cluster installation time.
```
" Provide a configuration option to allow use of alternate BGP implementations during
  MetalLB deployment."

markdgray commented on design/0001-frr.md:
```diff
@@ -0,0 +1,341 @@
+#### Story 2
+
+As a cluster administrator, I want to be able to specify a configuration option
+when deploying MetalLB in order to select the MetalLB BGP agent to be used by
+MetalLB
```
Fixed.

markdgray commented on design/speaker.md:
```diff
@@ -0,0 +1,128 @@
+BGP sessions are uniquely identified and configured by the parameters of the
+New() method which are currently:
+
+* **addr**: IP address and port of peer
```
updated to add this field.

markdgray commented on design/0001-frr.md:
```diff
@@ -0,0 +1,341 @@
+  specific FRR bugs without upgrading MetalLB. MetalLB will ship with a default
+  version but it will be possible to configure the version somehow (e.g. through
+  helm).
+* At the MetalLB layer you can restart speaker pods with no "global" issue.
```
It was copied verbatim from a conversation with @rata in a previous comment. However, I will reword slightly to remove the "global" comment as it is confusing.

markdgray commented on design/0001-frr.md:
```diff
@@ -0,0 +1,341 @@
+  issues with FRR.
+* This will not expose direct configuration of FRR to the MetalLB user.
+
+### Non-Goals
```
Ok, I have rephrased this, simplified it and also moved "This will not expose direct configuration of FRR to the MetalLB user." to non-goals as "Expose direct configuration of FRR to the MetalLB user."

markdgray commented on design/0001-frr.md:
```diff
@@ -0,0 +1,341 @@
+  interface that could be used to integrate other BGP stacks and will be a
+  template to do so.
+* This will not support runtime changing of BGP agent type.
+* This will not extend the functionality of the MetalLB BGP implementation in
+  in order to take advantage of the additional cababilities of FRR. This will
+  be a goal of future enhancements.
```
Rephrased

markdgray commented on design/0001-frr.md:
```diff
@@ -0,0 +1,341 @@
+authorized through a password. FRR does not encyrpt traffic over the TCP socket
+which may be required at a later stage.
+
+Initially, control of the FRR container will be achieved by sending messages to
```
I had a chat with some folks in Red Hat using FRR in Openstack. They suggested an approach that they take using the -c option to send a configuration file to FRR. They build this file using templates and gives the kind of declarative control that you are suggesting here. I will update this section.

markdgray replied:
I have updated this section stating configuration by a config file (which should also simplify the security discussion re TCP sockets, etc)

markdgray commented on design/0001-frr.md:
```diff
@@ -0,0 +1,341 @@
+  However, Layer 2 memberlist code will see a node leave and rejoin the cluster.
+  In BGP mode, these [things](https://metallb.universe.tf/concepts/bgp/#limitations)
+  can happen (in particular "Connection reset by peer") and can be mitigated.
+* Upgrade/downgrade will remove a node by unlabelling the node. At this point
```
This section was originally about potentially adding tests that upgrade MetalLB, making sure it works. However this section has now grown (through review comment additions) and I think it makes sense to move this section to "Design"

markdgray commented on design/0001-frr.md:
```diff
@@ -0,0 +1,341 @@
+FRR does not provide documented northbound Go bindings to allow configuration of
+BGP. There is an experimental gRPC interface. This interface may need to be
+productionized through the FRR community. When this has been satisfactorily
+achieved, we can start [Story 3](#story-3). Until that time, in order to
```
Removed that phrasing.

johananl commented on design/0001-frr.md:
```diff
@@ -0,0 +1,312 @@
+version but it will be possible to configure the version somehow (e.g. through
+Helm).
+
+You can restart speaker Pods for upgrades. However, Layer 2 memberlist code
```
```suggestion
Speaker Pods can be restarted for upgrades. However, Layer 2 memberlist code
```

"You" is inconsistent with the tone of the document.

johananl commented on design/0001-frr.md:
```diff
@@ -0,0 +1,312 @@
+cluster, there is no requirement that each node within the cluster maintains the
+same version or a compatible version of FRR. However, it is required that each
+version of FRR is compatible with the BGP peer with which it is peering. After
+the ‘speaker’ component is restarted the node can be labelled again.
```
```suggestion
the ‘speaker’ component is restarted, the node can be labelled again.
```

johananl commented on design/0001-frr.md:
```diff
@@ -0,0 +1,312 @@
+* As FRR has a larger code base and configuration space, it will have a larger
+  security attack surface relative to that of the current native BGP
+  implementation of MetalLB. This could be mitigated by hardening of the
+  configuration. For example, using [peer filtering](http://docs.frrouting.org/en/latest/bgp.html#peer-filtering)
```
```suggestion
  configuration, for example using [peer filtering](http://docs.frrouting.org/en/latest/bgp.html#peer-filtering)
```

johananl commented on design/0001-frr.md:
```diff
@@ -0,0 +1,312 @@
+  configured with a gRPC interface which can generate Go bindings.
+* Extensibility options: FRR provides additional extensibility via its `dplane`
+  and `FPM` interfaces. This may ease integration with Kubernetes networking
+  providers. There does not appear to be equivalents in BIRD which make it less
```
```suggestion
  providers. There seem to be no equivalents in BIRD, which makes BIRD less
```

johananl commented on design/0001-frr.md:
```diff
@@ -0,0 +1,312 @@
+  and `FPM` interfaces. This may ease integration with Kubernetes networking
+  providers. There does not appear to be equivalents in BIRD which make it less
+  extensible. This extensibility would enable FRR to integrate with dataplanes
+  other than the Linux kernel. For example, 1) hardware dataplanes 2) an Open
```
```suggestion
  other than the Linux kernel. Examples: 1) hardware dataplanes 2) an Open
```

johananl reviewed (approved):
Great work @markdgray. This is an easy read now and makes a lot of sense to me.

The remaining comments are all language-related nitpicks so I'm already LGTM-ing the PR and will leave it up to you to decide whether to fix them.

markdgray commented:
> Great work @markdgray. This is an easy read now and makes a lot of sense to me.
> 
> The remaining comments are all language-related nitpicks so I'm already LGTM-ing the PR and will leave it up to you to decide whether to fix them.

Thanks @johananl. All the changes seem reasonable. I have accepted your suggestions, squashed the commits and rebased against `main`. I think this is done now?

johananl commented:
LGTM @markdgray.
@daxmc99 @rata I'll merge this later today so last chance to weigh in :slightly_smiling_face:

fedepaol commented:
> At Ericsson we use the metallb `controller` but have developed an own `speaker` with the functions we need, for example;
> 
>     * IPv6
> 
>     * BFD
> 
>     * Static routes (with BFD)
> 
>     * Traffic separation (separate peers)
> 
> 
> At a meeting with `RedHat` we were asked to contribute with our speaker configuration as input to this PR, perhaps even as a definition of done;
> 
>     * [ECFE_Configuration_Examples.md](https://github.com/metallb/metallb/files/6631459/ECFE_Configuration_Examples.md)
> 
>     * [ECFE_Configuration_Examples.pdf](https://github.com/metallb/metallb/files/6631462/ECFE_Configuration_Examples.pdf)

@uablrek sorry for getting back on this so late.

What would be the use case for static BFD without BGP?

Your `static-bfd-peers` example states that the addressPool is still using BGP, but no BGP peers are configured. I assume this is to keep the example shorter, but it will still be use BGP. What is the use case for adding a BGP peer which is not the same as the BGP peer?

Thanks!

uablrek commented:
I have only passed the config so I might be wrong, but I think the metallb contoller requires a "protocol" entry so it is there but not used. Our setup uses the metallb controller with one modification; the yaml parsing is not "strict", so we can add things but must comply with existing syntax.

uablrek commented:
The remove-strict commit https://github.com/Nordix/metallb/commit/13b6cca8293817a8189e7b5ba71de0a2553b60ea

fedepaol commented:
> I have only passed the config so I might be wrong, but I think the metallb contoller requires a "protocol" entry so it is there but not used. Our setup uses the metallb controller with one modification; the yaml parsing is not "strict", so we can add things but must comply with existing syntax.

So, if I am getting your comment right, in the "static-bfd" context you are not using bgp at all (which kind of resonated with the "static-bfd" naming). In that case, how is the advertisement of the virtual ip performed?

uablrek commented:
> in the "static-bfd" context you are not using bgp at all

Right. The route is statically configured in the GW-router by the site owner. Sometimes the site owner does not allow routing protocols.

fedepaol commented:
> > in the "static-bfd" context you are not using bgp at all
> 
> Right. The route is statically configured in the GW-router by the site owner. Sometimes the site owner does not allow routing protocols.

So how does the GW-router node knows which node(s) own the ip address associated with the lb service in this case? Or to put in other words, where that static route is set towards?
Because that ip can move from one (or more) node to others.

uablrek commented:
> own the ip address associated with the lb service

No node "owns" the lb-address. But if `externalTrafficPolicy: local` is used only nodes where a server pods run can be used. I don't know if there is some clever use of BFD to _not_ respond on some nodes (similar to BGP). Let me ask around ...

fedepaol commented:
> > own the ip address associated with the lb service
> 
> No node "owns" the lb-address. But if `externalTrafficPolicy: local` is used only nodes where a server pods run can be used. I don't know if there is some clever use of BFD to _not_ respond on some nodes (similar to BGP). Let me ask around ...

On top of that, there's the bfd + node selector section where only a subset of nodes are enabled.

I'm still not sure I get how the routes are configured. Are those something like ecmp static rules using the nodes (all of them) as gateways for the vips or something like that?

champtar commented:
BFD is per node, BGP let you advertise and retract multiple IPs, so this static-bfd is only used for 1 IP I guess ?

mandydydy commented:
> > own the ip address associated with the lb service
> 
> No node "owns" the lb-address. But if `externalTrafficPolicy: local` is used only nodes where a server pods run can be used. I don't know if there is some clever use of BFD to _not_ respond on some nodes (similar to BGP). Let me ask around ...

static-bfd doesn't care about the externalTrafficPolicy. We do not have additional mechanism regarding `externalTrafficPolicy:local`. If users want to use `extrernalTrafficPolicy:local` with static-bfd, they need to configure the Gateway Router to forward the traffic to the desired worker nodes themselves.
static-bfd only takes down the static routes on Gateways Routers. So if the bfd session of a worker node is down, the traffic won't be forwarded to this worker node from the Gateway Routers.
